src/avram/nested_save_operation.cr

```diff
@@ -1,35 +1,61 @@
 module Avram::NestedSaveOperation
   macro has_one(type_declaration)
     {% name = type_declaration.var %}
-    {% type = type_declaration.type %}
-    @_{{ name }} : {{ type }} | Nil
+    {% type = type_declaration.type.resolve %}
+    {% model_type = type.superclass.superclass.type_vars.first.resolve %}
 
-    def {{ name }}
-      @_{{ name }} ||= {{ type }}.new(params)
-    end
+    {% assoc = T.resolve.constant(:ASSOCIATIONS).find do |assoc|
+         assoc[:relationship_type] == :has_one &&
+           assoc[:type].resolve.name == model_type.name
+       end %}
 
-    after_save save_nested_{{ name }}
+    {% unless assoc %}
+      {% raise "#{T} must have a has_one association with #{model_type}" %}
+    {% end %}
 
-    def save_nested_{{ name }}(record)
-      {{ name }}.{{ @type.constant(:FOREIGN_KEY).id }}.value = record.id
+    after_save save_{{ name }}
 
-      if !{{ name }}.save
+    def save_{{ name }}(saved_record)
+      unless {{ name }}.save
         mark_nested_save_operations_as_failed
         database.rollback
       end
     end
 
+    def {{ name }}
+      return update_{{ name }} unless new_record?
+
+      nested = create_{{ name }}
+
+      if persisted?
+        nested.{{ @type.constant(:FOREIGN_KEY).id }}.value = record.not_nil!.id
+      end
+
+      nested
+    end
+
     def nested_save_operations
       {% if @type.methods.map(&.name).includes?(:nested_save_operations.id) %}
         previous_def +
       {% end %}
       [{{ name }}]
     end
+
+    private def create_{{ name }}
+      @{{ name }} ||= {{ type }}.new(params)
```

Comment:
Let's move this memoization back up to the `def {{ name }}` and out of the create and update methods. You can memoize with multiple actions on the right side by doing
```crystal
@{{ name }} ||= begin
  # ...
end
```

Reply:
Sure. Will work on it.